pkg/scheduler/backend/queue/backoff_queue.go

```diff
@@ -0,0 +1,231 @@
+	bq := &backoffQueue{
+		clock:             clock,
+		podInitialBackoff: podInitialBackoffDuration,
+		podMaxBackoff:     podMaxBackoffDuration,
+	}
+	bq.podBackoffQ = heap.NewWithRecorder(podInfoKeyFunc, bq.podsCompareBackoffCompleted, metrics.NewBackoffPodsRecorder())
+	bq.podErrorBackoffQ = heap.NewWithRecorder(podInfoKeyFunc, bq.podsCompareBackoffCompleted, metrics.NewBackoffPodsRecorder())
+
+	return bq
+}
+
+func (bq *backoffQueue) podInitialBackoffDuration() time.Duration {
+	return bq.podInitialBackoff
+}
+
+func (bq *backoffQueue) podMaxBackoffDuration() time.Duration {
+	return bq.podMaxBackoff
+}
+
+// podsCompareBackoffCompleted is a less function of podBackoffQ and podErrorBackoffQ.
+func (bq *backoffQueue) podsCompareBackoffCompleted(pInfo1, pInfo2 *framework.QueuedPodInfo) bool {
+	bo1 := bq.getBackoffTime(pInfo1)
+	bo2 := bq.getBackoffTime(pInfo2)
+	return bo1.Before(bo2)
+}
+
+// shouldBeBackingoff returns true if a pod is still waiting for its backoff timer.
+// If this returns true, the pod should not be re-tried.
+func (bq *backoffQueue) shouldBeBackingoff(podInfo *framework.QueuedPodInfo) bool {
+	boTime := bq.getBackoffTime(podInfo)
+	return boTime.After(bq.clock.Now())
+}
+
+// getBackoffTime returns the time that podInfo completes backoff
+func (bq *backoffQueue) getBackoffTime(podInfo *framework.QueuedPodInfo) time.Time {
+	duration := bq.calculateBackoffDuration(podInfo)
+	backoffTime := podInfo.Timestamp.Add(duration)
+	return backoffTime
+}
+
+// calculateBackoffDuration is a helper function for calculating the backoffDuration
+// based on the number of attempts the pod has made.
+func (bq *backoffQueue) calculateBackoffDuration(podInfo *framework.QueuedPodInfo) time.Duration {
+	if podInfo.Attempts == 0 {
+		// When the Pod hasn't experienced any scheduling attempts,
+		// they aren't obliged to get a backoff penalty at all.
+		return 0
+	}
+
+	duration := bq.podInitialBackoff
+	for i := 1; i < podInfo.Attempts; i++ {
+		// Use subtraction instead of addition or multiplication to avoid overflow.
+		if duration > bq.podMaxBackoff-duration {
+			return bq.podMaxBackoff
+		}
+		duration += duration
+	}
+	return duration
+}
+
+func (bq *backoffQueue) forEachBackoffCompletedWithQueue(logger klog.Logger, fn func(pInfo *framework.QueuedPodInfo), queue *heap.Heap[*framework.QueuedPodInfo]) {
+	for {
+		pInfo, ok := queue.Peek()
+		if !ok || pInfo == nil {
+			break
+		}
+		pod := pInfo.Pod
+		if bq.shouldBeBackingoff(pInfo) {
+			break
+		}
+		_, err := queue.Pop()
+		if err != nil {
+			logger.Error(err, "Unable to pop pod from backoff queue despite backoff completion", "pod", klog.KObj(pod))
+			break
+		}
+		fn(pInfo)
+	}
+}
+
+// forEachBackoffCompleted runs fn for all pods from podBackoffQ and podErrorBackoffQ that completed backoff.
+func (bq *backoffQueue) forEachBackoffCompleted(logger klog.Logger, fn func(pInfo *framework.QueuedPodInfo)) {
```

Comment:
It's not a usual `forEach` because it takes pods out of the backoffQ(s), not just run a given function for each pod.

```suggestion
// forEachBackoffCompleted pops all pods from backoffQ that has completed backoff,
// and run fn with them.
func (bq *backoffQueue) popEachBackoffCompleted(logger klog.Logger, fn func(pInfo *framework.QueuedPodInfo)) {
```